File: Leetcode/0837 New 21 Game.ts
function new21Game(n: number, k: number, maxPts: number): number {
    if (k === 0) { 
        return 1;
    }
    const prob = new Array<number>(k).fill(0);
    let windowSum = Math.min(maxPts, n - k + 1);
    for (let i = k - 1; i >= 0; i--) {
        prob[i] = windowSum / maxPts;
        const exitIndex = i + maxPts; 
        let exitValue = 0;
        if (exitIndex < k) {
            exitValue = prob[exitIndex];
        } else if (exitIndex <= n) {
            exitValue = 1;
        }
        windowSum += prob[i] - exitValue;
    }
    return prob[0];
}

File: Leetcode/0837 New 21 Game.py
class Solution:
    def new21Game(self, n, k, max_pts):
        if k == 0:
            return 1
        prob = [0] * k
        window_sum = min(max_pts, n - k + 1)
        for i in range(k - 1, -1, -1):
            prob[i] = window_sum / max_pts
            exit_index, exit_value = i + max_pts, 0
            if exit_index < k:
                exit_value = prob[exit_index]
            elif exit_index <= n:
                exit_value = 1
            window_sum += prob[i] - exit_value
        return prob[0]
